### PR TITLE
Fix availability free-busy data pull, location change and status in CALDAV

### LINKED ISSUE #3589: Availability fetching is bugged for CalDAV
### Issue Summary

So, I found a weird one while trying to debug CalDAV (but I’m not sure if this one is limited to CalDAV or spreads across other integrations as well)

When we book a slot in one of the event types, only that slot within that specific event type is blocked, but other event types still have the slots during that booked time available. Note that the booking was made at the CalDAV calendar as well, yet I think the freebusy data fetched from it was empty.

### Steps to Reproduce

[Loom Video](https://www.loom.com/share/88df05c8a4224172a7b520bef4f4d2f4)

Any other relevant information. For example, why do you consider this a bug and what did you expect to happen instead?

### Technical details

Integration in use: CalDAV

### PR DESCRIPTION
**Pull request by alishaz-polymath**
## What does this PR do?

- Fixes the incompatibility between the `ICALstring` returned in the calendar objects by TSDAV's `fetchCalendarObjects` and the expected input from `ICAL.parse()`
- Adds a sanitizer for calendar object data retrieved by TSdav for further processing
- Fixes Location Change reflecting on the connected CalDAV
- Fixes Reschedule change reflecting on the connected CalDAV
- Fixes Freebusy data consideration to allow booking over events with 'TRANSPARENT' (free) status

Fixes #3589 #2988 #2236 #1588 

**Environment**: Staging(main branch)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by alishaz-polymath** [comment]
Testing further as the events created on the CalDAV integration calendar are not blocked still. Investigating...

File: packages/lib/CalendarService.ts [hunk 1]
```diff
@@ -251,12 +251,16 @@ export default abstract class BaseCalendarService implements Calendar {
     objects.forEach((object) => {
       if (object.data == null) return;
 
-      const jcalData = ICAL.parse(object.data);
+      const jcalData = ICAL.parse(sanitizeCalendarObject(object));
```
Comment (alishaz-polymath) [inline packages/lib/CalendarService.ts:254]: Depending on the CalDAV service provider, the `object.data` may or may not have `\r` or `\r\n` as the EOL. `ICAL.parse` expects it to have '\r\n' to work as expected. Also, the `object.data` received from the `fetchCalendarObjects` function is at times faulty in placing the EOL, which is why we need to make sure that it doesn't break the structure of expected input for `ICAL.parse`, hence the need for sanitization.

File: packages/lib/CalendarService.ts [hunk 2]
```diff
@@ -251,12 +251,16 @@ export default abstract class BaseCalendarService implements Calendar {
...
+
       const vcalendar = new ICAL.Component(jcalData);
       const vevent = vcalendar.getFirstSubcomponent("vevent");
       const event = new ICAL.Event(vevent);
       const vtimezone = vcalendar.getFirstSubcomponent("vtimezone");
 
+      // if event status is free or transparent, return
+      if (vevent?.getFirstPropertyValue("transp") === "TRANSPARENT") return;
```
Comment (alishaz-polymath) [inline packages/lib/CalendarService.ts:262]: If an event is added as a Free event, meaning we should allow booking over it, this is how we identify the free state and ignore any further processing

File: packages/lib/sanitizeCalendarObject.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { DAVObject } from "tsdav";
+
+export const sanitizeCalendarObject = (obj: DAVObject) => {
+  return obj.data
+    .replaceAll("\r\n", "\r")
+    .replaceAll("\r", "\r\n")
+    .replaceAll(/(: \r\n|:\r\n|\r\n:|\r\n :)/gm, ":")
+    .replaceAll(/(; \r\n|;\r\n|\r\n;|\r\n ;)/gm, ";")
+    .replaceAll(/(= \r\n|=\r\n|\r\n=|\r\n =)/gm, "=");
```
Comment (alishaz-polymath) [inline packages/lib/sanitizeCalendarObject.ts:9]: We make sure the `ICAL.parse` works with the output of this return

**Review by alishaz-polymath — COMMENTED** [review]
Self Review Added

**Comment by PeerRich** [comment]
whats the best way for us to test it? if this doesn't touch anything core we can merge it and also test with users who have more diverse CalDav services

**Comment by alishaz-polymath** [comment]
> whats the best way for us to test it? if this doesn't touch anything core we can merge it and also test with users who have more diverse CalDav services

The best way to test this would be across multiple CalDAV services and unfortunately there’s no easy or quick way of doing that without the merge. This doesn’t touch anything core, only the CalDAV calendar manager.

**Comment by alishaz-polymath** [comment]
Closed in favor of #3958

**Pull request closed without merging**